ilpy/expressions.py

```diff
@@ -0,0 +1,360 @@
+
+
+class UnaryOp(Expression, ast.UnaryOp):
+    """A unary operation.
+
+    `op` is the operator, and `operand` any expression node.
+    """
+
+    def __init__(self, op: ast.unaryop, operand: Expression, **kwargs: Any) -> None:
+        super().__init__(op, Expression._cast(operand), **kwargs)
+
+
+class Constant(Expression, ast.Constant):
+    """A constant value.
+
+    The `value` attribute contains the Python object it represents.
+    types supported: int, float
+    """
+
+    def __init__(
+        self, value: float | int, kind: str | None = None, **kwargs: Any
+    ) -> None:
+        if not isinstance(value, (float, int)):
+            raise TypeError("Constants must be numbers")
+        super().__init__(value, kind, **kwargs)
+
+
+class Variable(Expression, ast.Name):
+    """A variable.
+
+    `id` holds the index as a string (becuase ast.Name requires a string).
+
+    The special attribute `index` is added here for the purpose of storing
+    the index of a variable in a solver's variable list: ``Variable('u', index=0)``
+    """
+
+    def __init__(self, id: str, index: int | None = None) -> None:
+        self.index = index
+        super().__init__(str(id), ctx=ast.Load())
+
+    def __hash__(self) -> int:
+        # allow use as dict key
+        return id(self)
+
+
+# conversion between ast comparison operators and ilpy relations
+# TODO: support more less/greater than operators
+OPERATOR_MAP: dict[type[ast.cmpop], Relation] = {
+    ast.LtE: Relation.LessEqual,
+    ast.Eq: Relation.Equal,
+    ast.GtE: Relation.GreaterEqual,
+}
+
+
+def _get_relation(expr: Expression) -> Relation | None:
+    seen_compare = False
+    relation: Relation | None = None
+    for sub in ast.walk(expr):
+        if isinstance(sub, Compare):
+            if seen_compare:
+                raise ValueError("Only single comparisons are supported")
+
+            op_type = type(sub.ops[0])
+            try:
+                relation = OPERATOR_MAP[op_type]
+            except KeyError as e:
+                raise ValueError(f"Unsupported comparison operator: {op_type}") from e
+            seen_compare = True
+    return relation
+
+
+def _expression_to_constraint(expr: Expression) -> LinearConstraint:
+    """Convert an expression to a `LinearConstraint`."""
+    constraint = LinearConstraint()
+    if relation := _get_relation(expr):
+        constraint.set_relation(relation)
+    for var, coefficient in _get_coefficients(expr).items():
+        if var is None:
+            # None is the constant term, which is the right hand side of the
+            # comparison.
+            constraint.set_value(-coefficient)
+        elif coefficient != 0:
+            if var.index is None:
+                raise ValueError(
+                    "All variables in a constraint expression must have an index"
+                )
+            constraint.set_coefficient(var.index, coefficient)
+    return constraint
+
+
+def _get_coefficients(
```

Comment:
would appreciate a careful look at this function... which is responsible for reducing an expression to coefficients (even if a term appears multiple times or is nested inside parentheses)

Reply:
Looks good to me!